### PR TITLE
Added sonar and vision sensors

### PR DESCRIPTION
**Pull request by WilliamBallbach**
Replaced remoteSensors.py with separate files for sonar and vision. I'm not sure if everything is done correctly and/or completely, but at least most of the necessary code should be there.

### CODE REVIEW

**Comment by WilliamBallbach** [comment]
The sonar, vision, and robot.py currently use ports for testing, not the real ports

**Comment by mcolinj** [comment]
We should probably try to test this code on the robot (using the proper addresses) and some kind of vision and sonar before we integrate it into master.     However, we should invite @cmonk to use it for his vision testing work.     Maybe you can be around when he is testing to provide some support as needed.    For now, you can keep pulling in updates from master so you are up-to-date, and so the final version will merge easily.

File: robot.py [hunk 1]
```diff
@@ -129,6 +129,12 @@ def teleopInit(self):
         self.pistons_activated = False
         self.forward = 0
 
+        #Vision sensor
+        self.visionSensor = VisionSensor('127.0.0.1', 8812)
```
Comment (mcolinj) [inline robot.py]: Can we figure out how to determine if we are "doing simulator" and have this use the preferred IP if we are not?    I don't we want to commit this loopback into master.

File: robot.py [hunk 2]
```diff
@@ -129,6 +129,12 @@ def teleopInit(self):
...
+
+        #Sonar sensor
+        self.sonarSensor = SonarSensor('127.0.0.1', 8813)
```
Comment (mcolinj) [inline robot.py]: Same issue here, no loopback to be committed to master.

Perhaps we could having something near the top:

if we_are_doing_simulation:
    SONAR_IP = '127.0.0.1'
else:
   SONAR_IP = '10.10.76.9'

Then we can keep this part of the code nice and clean.
While you are at, give the ports nice ALL_CAPS names as well.

Even if we can't figure out how to do that simulation thing, we could have our own
function  doing_simulation() that return False (unless we hack it...)

File: subsystems/sonarSensor.py [hunk 3]
```diff
@@ -1,9 +1,6 @@
 import wpilib
 import json
-try:
```
Comment (mcolinj) [inline subsystems/sonarSensor.py]: We definitely need a plan of action for this.   I don't think we have the udp_channel library pip installed on the robot.    It is possible, but we'll need know what version and start testing with it regularly.

File: subsystems/sonarSensor.py [hunk 4]
```diff
@@ -33,7 +31,7 @@ def createChannel(self):
         sonar_ip = self.sonar_ip
         sonar_port = self.sonar_port
         try:
-            self.channel = UDPChannel(local_ip="10.10.76.2", local_port=5811,
```
Comment (mcolinj) [inline subsystems/sonarSensor.py]: We could be clever here and check if the sonar_ip is '127.0.0.1', and then just use that for the local_ip as well, otherwise the robot IP.  That way this works two interesting ways.

File: subsystems/visionSensor.py [hunk 5]
```diff
@@ -44,7 +42,7 @@ def createChannel(self):
         vision_ip = self.vision_ip
         vision_port = self.vision_port
         try:
-            self.channel = UDPChannel(local_ip="10.10.76.2", local_port=5880,
```
Comment (mcolinj) [inline subsystems/visionSensor.py]: We also need to explicitly test to make sure this failure mode will leave a None in the self.channel.    Might need to explicitly set it to None if the UDPChannel creation throws an exception.  (and maybe it should).    Not certain otherwise how to recover from a failure to create something.

**Comment by mcolinj** [comment]
This is shaping up nicely.   I'd like you to do one more thing   

We need to get the udp_channel into the mix before this collapses to master, otherwise everyone's everything is going to break.        Not sure what you prefer.   We could add the file to the repo and work from there (installing it into a library in at season's end),   or, we could get the latest changes incorporated into the library, update the requirements.txt, send a note out to everyone to install the new library, and then commit this relying on the library.

thanks,

matthew

File: subsystems/sonarSensor.py [hunk 6]
```diff
@@ -11,10 +11,11 @@ class SonarSensor:
     Example:
         sonarSensor = SonarSensor('10.10.76.11', 5811)
     """
-    def __init__(self, sensor_ip, listen_port, logger=None):
+    def __init__(self, sensor_ip, listen_port, simulation=False, logger=None):
```
Comment (mcolinj) [inline subsystems/sonarSensor.py]: One more suggestion.    I think because you've put the sim/nosim switch inside, you can do the check in the constructor instead of in the createChannel function.  Just check, save the appropriate values in the instance variables, and then just use the self.ip and self.port in the createChannel().

File: subsystems/visionSensor.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            VISION_IP = self.vision_ip
+            VISION_PORT = self.vision_port
+            LOCAL_IP = "10.10.76.2"
+        else:
+            VISION_IP = "127.0.0.1"
+            VISION_PORT = 8812
+            LOCAL_IP = "127.0.0.1"
+        
+        try:
+            channel = UDPChannel(local_ip=LOCAL_IP, local_port=VISION_PORT,
```
Comment (mcolinj) [inline subsystems/visionSensor.py]: This needs to be self.channel

File: subsystems/visionSensor.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        else:
+            VISION_IP = "127.0.0.1"
+            VISION_PORT = 8812
+            LOCAL_IP = "127.0.0.1"
+        
+        try:
+            channel = UDPChannel(local_ip=LOCAL_IP, local_port=VISION_PORT, 
+                                      remote_ip=VISION_IP, remote_port=VISION_PORT)
+        except:
+            channel = None
```
Comment (mcolinj) [inline subsystems/visionSensor.py]: Same here.

File: subsystems/visionSensor.py (hunk 7 above)
Comment (mcolinj) [reply subsystems/visionSensor.py]: Nope.  I got that wrong.   This is local variable to return.  By Bad.  Resolved.
